rasterio/_io.pyx

```diff
@@ -1145,10 +1145,12 @@ cdef class MemoryFileBase:
         cdef VSILFILE *fp = NULL
 
         if file_or_bytes:
-            if hasattr(file_or_bytes, 'read'):
+            if hasattr(file_or_bytes, "read"):
                 initial_bytes = file_or_bytes.read()
             elif isinstance(file_or_bytes, bytes):
                 initial_bytes = file_or_bytes
+            elif hasattr(file_or_bytes, "itemsize"):
+                initial_bytes = bytes(file_or_bytes)
```

Comment:
Is this for numpy arrays?

Reply:
It's to allow MemoryFile cloning like `new = MemoryFile(old.getbuffer())`. It's not something one would do in practice but covers some existing edge cases in our tests.